src/nimble_uq.cc

```diff
@@ -0,0 +1,790 @@
+  //              1.0 ==> nominal + range
+  //              e.g., if distribution is Gaussian, range=3*sigma contains 99.7% of samples
+
+  int ns = GetNumSamples();
+
+  for (int np = 0; np < nparameters_; np++) {
+    double slope = ranges_[np];
+    double shift = nominal_parameter_values_[np];
+
+    for (int n = 0; n < ns; n++) {
+      parameter_samples_[n][np] *= slope;
+      parameter_samples_[n][np] += shift;
+    }
+  }
+}
+//===========================================================================
+void
+UqModel::Initialize()
+{
+  if (samples_from_file_) {
+    ReadSamples();
+    ScaleParameters();  // PARMETERS IN FILE ARE in -1.0 : 1.0 RANGE
+  } else {
+    SampleParameters();  // NEED TO DO THIS ONLY IF THEY HAVENT BEEN READ FROM FILE
+    CenterParameters();
+    WriteSamples();
+  }
+  Allocate();
+  InitializeAnalyses();
+  initialized_ = true;
+}
+//===========================================================================
+void
+UqModel::Finalize()
+{
+}
+
+//===========================================================================
+void
+UqModel::Allocate()
+{
+  int n = GetNumSamples();
+  int offnominal_displacement_ids[n],  // tmp?
+      offnominal_internal_force_ids[n], offnominal_velocity_ids[n];
+  for (int nuq = 0; nuq < n; nuq++) {
+    offnominal_displacement_ids[nuq] =
+        data_->AllocateNodeData(nimble::VECTOR, "off_nom_displacement_" + std::to_string(nuq), nnodes_);
+    offnominal_velocity_ids[nuq] =
+        data_->AllocateNodeData(nimble::VECTOR, "off_nom_velocity_" + std::to_string(nuq), nnodes_);
+    offnominal_internal_force_ids[nuq] =
+        data_->AllocateNodeData(nimble::VECTOR, "off_nom_internal_force_" + std::to_string(nuq), nnodes_);
+  }
+  // for time averages
+  int sum_displacement_id  = data_->AllocateNodeData(nimble::VECTOR, "sum_displacement", nnodes_);
+  int sum2_displacement_id = data_->AllocateNodeData(nimble::VECTOR, "sum2_displacement", nnodes_);
+  int min_displacement_id  = data_->AllocateNodeData(nimble::VECTOR, "min_displacement", nnodes_);
+  int max_displacement_id  = data_->AllocateNodeData(nimble::VECTOR, "max_displacement", nnodes_);
+}
+//===========================================================================
+void
+UqModel::Setup()
+{
+  if (!initialized_) { return; }
+  mass_ = data_->GetNodeData(data_->GetFieldId("lumped_mass"));
+  offnominal_displacements_.clear();
+  offnominal_velocities_.clear();
+  offnominal_forces_.clear();
+  for (int nuq = 0; nuq < nsamples_; nuq++) {
+    offnominal_displacements_.push_back(
+        data_->GetNodeData(data_->GetFieldId("off_nom_displacement_" + std::to_string(nuq))));
+    offnominal_velocities_.push_back(data_->GetNodeData(data_->GetFieldId("off_nom_velocity_" + std::to_string(nuq))));
+    offnominal_forces_.push_back(
+        data_->GetNodeData(data_->GetFieldId("off_nom_internal_force_" + std::to_string(nuq))));
+  }
+}
+//===========================================================================
+void
```

Comment:
these parts of the verlet integrator can be moved into the uq block now that it see the off_nom data